# Industrial Energy Analytics (Spark Capstone)
**Sector:** Infraestructura Eléctrica e Ingeniería de Datos  
**Responsable:** Aitor (Ingeniero Técnico Industrial)  
**Fecha:** Febrero 2026

## Indice

## Configuración del Entorno (Collapsed)

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, count, when, isnan, isnull

# Inicializamos Spark con optimización de memoria
spark = SparkSession.builder \
    .appName("IndustrialEnergyAnalytics") \
    .config("spark.driver.memory", "16g") \
    .config("spark.executor.memory", "8g") \
    .getOrCreate()

# Verificamos la sesión
spark

## Ingesta y Profiling Inicial (Data Diagnosis)

In [2]:
# Definimos la ruta hacia la carpeta source
path = "../data_storage/source/household_power_consumption.txt"

# Cargamos el archivo con las opciones correctas para este dataset
df_clean = spark.read.options(header='True', sep=';', inferSchema='True', naStrings='?').csv(path)

# 1. Verificación del KPI de volumen
print(f"Total de registros cargados: {df_clean.count():,}")

# 2. Análisis del esquema detectado
print("\nEstructura detectada (Esquema):")
df_clean.printSchema()

# 3. Muestra de datos para inspección visual
print("\nMuestra de los primeros 5 registros:")
df_clean.show(5)

Total de registros cargados: 2,075,259

Estructura detectada (Esquema):
root
 |-- Date: string (nullable = true)
 |-- Time: timestamp (nullable = true)
 |-- Global_active_power: string (nullable = true)
 |-- Global_reactive_power: string (nullable = true)
 |-- Voltage: string (nullable = true)
 |-- Global_intensity: string (nullable = true)
 |-- Sub_metering_1: string (nullable = true)
 |-- Sub_metering_2: string (nullable = true)
 |-- Sub_metering_3: double (nullable = true)


Muestra de los primeros 5 registros:
+----------+-------------------+-------------------+---------------------+-------+----------------+--------------+--------------+--------------+
|      Date|               Time|Global_active_power|Global_reactive_power|Voltage|Global_intensity|Sub_metering_1|Sub_metering_2|Sub_metering_3|
+----------+-------------------+-------------------+---------------------+-------+----------------+--------------+--------------+--------------+
|16/12/2006|2026-02-25 17:24:00|             

## Data Wrangling e Ingeniería de Atributos (Feature Engineering)

Para optimizar el rendimiento en el entorno local (MEDION i5, 32GB), se ha diseñado un pipeline de **Data Wrangling** que unifica la limpieza de datos y la creación de variables derivadas en un único flujo de trabajo de Spark. Este proceso es fundamental para transformar registros brutos de telemetría en dimensiones con valor para la auditoría energética de una vivienda de gran escala.

### 1. Saneamiento de Telemetría (Data Cleansing - Soporte H3)
* **Acción:** Identificación y reemplazo del carácter `'?'` (lecturas fallidas del sensor) por valores `null`, seguido de una purga mediante `na.drop()`.
* **Justificación Técnica:** El diagnóstico de **Estabilidad de Tensión (H3)** requiere precisión matemática absoluta. Eliminar estas lagunas de información evita sesgos en el cálculo de caídas de tensión y asegura que las correlaciones entre la potencia demandada y el voltaje sean fiables para prevenir la fatiga de la electrónica de la vivienda.

### 2. Reconstrucción del Eje Cronológico (Data Transformation - Soporte H1)
* **Acción:** Unificación de las columnas `Date` y `Time` en un objeto `Full_Timestamp` de alta resolución.
* **Justificación Técnica:** El análisis de **Tasa de Simultaneidad (H1)** depende de la capacidad de secuenciar eventos. Sin este eje cronológico, sería imposible detectar los picos de demanda máxima que ocurren cuando coinciden grandes cargas (climatización + sistemas de bombeo), lo cual es crítico para optimizar el término de potencia coratada.

### 3. Ingeniería de Atributos de Hábitos (Feature Engineering - Soporte H2)
* **Acción:** Extracción de las dimensiones `Hour`, `Day_Number` e `Is_Weekend`.
* **Justificación Técnica:** Para validar la hipótesis de **Consumo Residual (H2)**, es imperativo segmentar los datos según las rutinas de la vivienda. 
    * La columna `Hour` permite aislar el consumo de madrugada.
    * `Is_Weekend` permite comparar el comportamiento de la vivienda en días de descanso frente a laborables.
    * Estas variables actúan como "filtros inteligentes" para cuantificar el ahorro potencial mediante la optimización de sistemas de domótica y protocolos de eficiencia pasiva


In [3]:
from pyspark.sql.functions import col, concat, lit, to_timestamp, dayofweek, hour, when

# 1. TRATAMIENTO DE VALORES FALTANTES (Limpieza)
df_pre_clean = df_clean.replace('?', None)
df_no_nulls = df_pre_clean.na.drop()

# 2. ENRIQUECIMIENTO DE DATOS (Feature Engineering Integrado)
df_final = df_no_nulls.withColumn(
    # Generación de eje temporal para análisis de picos (H1)
    "Full_Timestamp", 
    to_timestamp(concat(col("Date"), lit(" "), col("Time").cast("string").substr(12, 8)), "d/M/y HH:mm:ss")
).withColumn(
    # Dimensión horaria para detección de consumos nocturnos (H2)
    "Hour", hour(col("Full_Timestamp"))
).withColumn(
    # Dimensión diaria para análisis de rutinas semanales
    "Day_Number", dayofweek(col("Full_Timestamp")) 
).withColumn(
    # Etiquetado de fines de semana para auditoría de inactividad (H2)
    "Is_Weekend", when(col("Day_Number").isin(1, 7), True).otherwise(False)
)

# 3. VERIFICACIÓN DE SALIDA
print(f"Registros listos para análisis: {df_final.count():,}")
df_final.select("Full_Timestamp", "Hour", "Day_Number", "Is_Weekend").show(5)


Registros listos para análisis: 2,049,280
+-------------------+----+----------+----------+
|     Full_Timestamp|Hour|Day_Number|Is_Weekend|
+-------------------+----+----------+----------+
|2006-12-16 17:24:00|  17|         7|      true|
|2006-12-16 17:25:00|  17|         7|      true|
|2006-12-16 17:26:00|  17|         7|      true|
|2006-12-16 17:27:00|  17|         7|      true|
|2006-12-16 17:28:00|  17|         7|      true|
+-------------------+----+----------+----------+
only showing top 5 rows



## Abstracción a Capa SQL
* **Acción:** Creación de la vista temporal `energy_data`.
* **Justificación:** Separamos la lógica de preparación (Spark/Python) de la lógica de negocio (SQL). Esto permite realizar consultas complejas sobre las métricas industriales con la eficiencia del motor de ejecución de Spark.

In [4]:
# REGISTRO DE VISTA PARA SQL
# Justificación: Registramos 'energy_data' para poder validar las hipótesis H1, H2 y H3 
# usando lenguaje SQL, facilitando agrupaciones complejas y filtros de ingeniería.
df_final.createOrReplaceTempView("power_data")

## Validación de Hipótesis de Ahorro Energético

En esta sección, utilizaremos **Spark SQL** para interrogar el dataset y validar tres hipótesis clave que impactan directamente en la factura eléctrica.

## Hipótesis 1: (Optimización de la Curva de Carga y Simultaneidad)
* **Definicion:** La instalación registra picos de demanda máxima vinculados a una elevada **tasa de simultaneidad** de equipos de gran potencia (Sub_metering_3 climatización).
* **Objetivo:** Identificar las ventanas temporales de máxima demanda donde la coincidencia de cargas críticas (climatización, bombeo y térmicos) dispara la potencia activa.
* **Valor de Negocio:** En instalaciones monofásicas de alta capacidad, el pico máximo registrado define la potencia contratada necesaria. Mediante la detección de estos eventos, es posible proponer un **escalonamiento de cargas** que reduzca el término fijo de la factura sin comprometer el confort o la operatividad de la vivienda.

In [5]:
import time
# H1.1:Perfil de Carga Horario
# Nota: Si AVG es bajo pero el MAX es altísimo, hay un problema de "pico puntual" (simultaneidad). Si ambos son altos, es un consumo base elevado.

start_h1 = time.time()

picos_horarios = spark.sql("""

SELECT 
    Hour as hora_del_dia,
    ROUND(AVG(Global_active_power), 3) as potencia_promedio_kw,
    ROUND(MAX(Global_active_power), 3) as potencia_maxima_kw
FROM 
    power_data
GROUP BY 
    Hour
ORDER BY 
    hora_del_dia ASC;

""")
picos_horarios.show(24)
print(f"H1 procesada en: {time.time() - start_h1:.2f} segundos")

+------------+--------------------+------------------+
|hora_del_dia|potencia_promedio_kw|potencia_maxima_kw|
+------------+--------------------+------------------+
|           0|               0.659|             7.698|
|           1|               0.539|             9.938|
|           2|               0.481|             6.248|
|           3|               0.445|             4.426|
|           4|               0.444|             4.796|
|           5|               0.454|             6.376|
|           6|               0.792|              8.31|
|           7|               1.502|             9.486|
|           8|               1.461|             8.126|
|           9|               1.332|             7.732|
|          10|               1.261|             9.078|
|          11|               1.246|             8.504|
|          12|               1.207|             9.224|
|          13|               1.145|              9.59|
|          14|               1.083|             8.418|
|         

**Análisis Técnico H1.1:**  
Se confirma una infrautilización de la potencia contratada. Los resultados de la H1.1 muestran una gran dispersión entre el AVG y el MAX, especialmente a la 01:00 y entre las 19:00-21:00. Esto valida la sospecha de picos por simultaneidad.

---

In [6]:
# H1.2:Identificación de la "Ventana Crítica"
# Nota: Si en este Top 20 la hora siempre es XX:00 o XX:30, es un automatismo (programación). Si es aleatoria, es comportamiento humano.

start_h1 = time.time()

ventana_critica = spark.sql("""

SELECT 
    Full_Timestamp as fecha_hora,
    Global_active_power as potencia_kw,
    -- Añadimos el día de la semana para ver patrones (1=Domingo, 2=Lunes...)
    Day_Number as dia_semana 
FROM 
    power_data
ORDER BY 
    Global_active_power DESC
LIMIT 20;

""")
ventana_critica.show(20)
print(f"H1 procesada en: {time.time() - start_h1:.2f} segundos")

+-------------------+-----------+----------+
|         fecha_hora|potencia_kw|dia_semana|
+-------------------+-----------+----------+
|2008-11-30 20:18:00|      9.994|         1|
|2008-10-19 01:27:00|      9.938|         1|
|2007-03-04 19:35:00|      9.922|         1|
|2009-11-24 19:37:00|      9.732|         3|
|2009-11-24 19:36:00|      9.726|         3|
|2009-11-24 19:38:00|      9.726|         3|
|2010-11-20 18:22:00|      9.724|         7|
|2009-11-24 19:34:00|      9.718|         3|
|2009-11-24 19:35:00|      9.714|         3|
|2009-11-24 19:39:00|      9.708|         3|
|2009-02-22 17:10:00|      9.694|         1|
|2010-11-20 18:26:00|      9.694|         7|
|2007-12-23 19:29:00|      9.686|         1|
|2008-01-26 19:32:00|      9.666|         7|
|2010-11-20 18:23:00|      9.656|         7|
|2008-01-26 19:54:00|      9.652|         7|
|2010-11-20 18:25:00|      9.648|         7|
|2010-11-20 18:24:00|      9.630|         7|
|2008-01-26 19:33:00|      9.628|         7|
|2008-01-2

**Análisis Técnico H1.2:**  
* Clúster de estrés: Concentración anómala de picos el 24/11/2009 (19:30h), indicando simultaneidad crítica.
* Patrón temporal: Predominio de máximos en fines de semana (Días 1 y 7) y franja de tarde-noche.
* Origen del consumo: La variabilidad en los minutos de los picos confirma un patrón de comportamiento humano y descarta arranques programados o automatismos fijos.
* Límite: La demanda satura sistemáticamente en el umbral de los 10 kW.

---

In [7]:
# H1.3:Análisis de Simultaneidad (Cruce de Sub-medidores)
# Nota: Si pct_clima (Sub 3) es el 80% del pico, tu hipótesis se confirma: el sistema de climatizacion es el culpables.


start_h1 = time.time()

picos_criticos = spark.sql("""

WITH PicosCriticos AS (
    -- Primero filtramos solo los momentos de estrés real (> 8 kW)
    SELECT * FROM power_data
    WHERE Global_active_power > 8
)
SELECT 
    Full_Timestamp as fecha_hora,
    Global_active_power as total_kw,
    
    -- Calculamos la potencia aprox de cada sub-medidor en kW
    ROUND((Sub_metering_1 * 0.06),2) as cocina_kw,
    ROUND((Sub_metering_2 * 0.06),2) as lavadero_kw,
    ROUND((Sub_metering_3 * 0.06),2) as clima_agua_kw,
    
    -- Calculamos el porcentaje de "culpa" de cada uno
    ROUND(((Sub_metering_1 * 0.06) / Global_active_power) * 100, 2) as pct_cocina,
    ROUND(((Sub_metering_2 * 0.06) / Global_active_power) * 100, 2) as pct_lavadero,
    ROUND(((Sub_metering_3 * 0.06) / Global_active_power) * 100, 2) as pct_clima,
    
    -- Lo que sobra es consumo "invisible" (luces, TV, ordenador, pérdidas)
    ROUND(
        (Global_active_power - ((Sub_metering_1 + Sub_metering_2 + Sub_metering_3) * 0.06)) 
        / Global_active_power * 100, 2
    ) as pct_otros
FROM 
    PicosCriticos
ORDER BY 
    Global_active_power DESC;

""")
picos_criticos.show(20)
print(f"H1 procesada en: {time.time() - start_h1:.2f} segundos")

+-------------------+--------+---------+-----------+-------------+----------+------------+---------+---------+
|         fecha_hora|total_kw|cocina_kw|lavadero_kw|clima_agua_kw|pct_cocina|pct_lavadero|pct_clima|pct_otros|
+-------------------+--------+---------+-----------+-------------+----------+------------+---------+---------+
|2008-11-30 20:18:00|   9.994|     2.16|        4.2|         1.02|     21.61|       42.03|    10.21|    26.16|
|2008-10-19 01:27:00|   9.938|      2.1|       4.02|         1.02|     21.13|       40.45|    10.26|    28.15|
|2007-03-04 19:35:00|   9.922|     2.04|        4.2|         0.96|     20.56|       42.33|     9.68|    27.43|
|2009-11-24 19:37:00|   9.732|     2.16|       4.14|         1.02|     22.19|       42.54|    10.48|    24.78|
|2009-11-24 19:36:00|   9.726|     2.16|        4.2|         1.02|     22.21|       43.18|    10.49|    24.12|
|2009-11-24 19:38:00|   9.726|     2.16|        4.2|         1.02|     22.21|       43.18|    10.49|    24.12|
|

**Análisis Técnico H1.3:**  
* Identificación de Cargas: El Sub-medidor 2 (Lavadero) es el principal driver de los picos, contribuyendo con más del 42% de la demanda total en momentos críticos.
* Interacción Crítica: Se confirma que los picos de ~10 kW se producen por la coincidencia del Lavadero con la Cocina (22%) y un consumo base elevado (25%).
* Refutación Parcial: Contrario a la hipótesis inicial, el sistema de climatización (Sub 3) tiene un impacto minoritario (<11%) en los eventos de máximo estrés analizados.ados.

---

In [8]:
# H1.4: Cálculo de la Tasa de Utilización (Factor de Carga)


start_h1 = time.time()

Factor_de_Carga = spark.sql("""

SELECT 
    ROUND(AVG(Global_active_power), 3) as potencia_media_sistema,
    MAX(Global_active_power) as potencia_pico_sistema,
    
    -- Factor de Carga (0 a 1). Cuanto más bajo, más ineficiente es la instalación
    -- (tienes mucha potencia contratada para usarla muy poco tiempo)
    ROUND(AVG(Global_active_power) / MAX(Global_active_power), 3) as factor_de_carga
FROM 
    power_data;

""")
Factor_de_Carga.show(20)
print(f"H1 procesada en: {time.time() - start_h1:.2f} segundos")

+----------------------+---------------------+---------------+
|potencia_media_sistema|potencia_pico_sistema|factor_de_carga|
+----------------------+---------------------+---------------+
|                 1.092|                9.994|          0.109|
+----------------------+---------------------+---------------+

H1 procesada en: 1.00 segundos


**Análisis Técnico H1.4:**  
* Factor de Carga Crítico: El valor de 0.1092 confirma una instalación extremadamente ineficiente. El sistema está dimensionado para eventos de pico muy poco frecuentes.
* Infrautilización: La potencia promedio representa solo el 10.9% de la capacidad máxima demandada, lo que indica un margen masivo para la optimización financiera mediante la gestión de la demanda.

---

## Conclusiones del Análisis: Hipótesis 1 (H1)

### **Estado de Validación: ✅ HIPÓTESIS VALIDADA**

Tras el procesamiento de grandes volúmenes de datos con Spark SQL y el análisis de los registros de demanda máxima (Top 20), se extraen las siguientes conclusiones técnicas:

* **Identificación de Cargas Críticas:** Se refuta la sospecha inicial sobre la climatización (Sub-3). El estrés de la instalación es provocado por la **simultaneidad entre la Cocina (Sub-1) y el Lavadero (Sub-2)**, que juntos representan más del **65% de la demanda total** en eventos críticos superiores a los 9 kW.
* **Caracterización del Consumo:** La aleatoriedad en los minutos de registro de los picos confirma un **patrón de comportamiento humano** vinculado a hábitos domésticos, descartando arranques programados o automatismos fijos. Los eventos de máximo estrés se concentran de forma recurrente en **fines de semana** y en la franja horaria de **19:00 a 21:00**.
* **Diagnóstico de Eficiencia (Load Factor):** Se ha obtenido un factor de carga d **0.1092**. Este valor indica una eficiencia extremadamente baja (10.9%), confirmando que la instalación está dimensionada para cubrir picos de demanda que son **9.1 veces superiores** a la potencia media utilizada ($1.09$ kW).
* **Tratamiento de Anomalías:** Se han identificado eventos aislados de gran magnitud (ej. picos a las **01:00 AM** o el clúster del **24/11/2009**) con una desviación estándar muy elevada respecto al promedio horario ($0.53$ kW vs $9.9$ kW). Al no responder a un patrón recurrente, estos registros se clasifican como *outliers* y se descartan para el dimensionamiento de la potencia contratada.


### **Recomendaciones Estratégicas e Impacto en Negocio**

Para transformar estos *insights* en ahorro económico real, se proponen las siguientes acciones:

1.  **Reingeniería de Hábitos (Escalonamiento de Cargas):** Se debe evitar el uso simultáneo del área de lavandería (lavadora/secadora) durante la ventana crítica de cocina (19:00 - 21:00). Desplazar el consumo del Sub-medidor 2 fuera de esta franja reduciría los picos de ~10 kW a un entorno de **5.5 - 6.0 kW** de forma segura.
2.  **Optimización del Suministro:** Basándonos en la estabilización de la curva de demanda mediante el escalonamiento, se recomienda **reducir la potencia contratada de los 9.2 kW actuales a 5.75 kW** (escalón normalizado).
3.  **Impacto Financiero:** Esta reducción del **37.5% en la potencia contratada** supone un ahorro directo y permanente en el término fijo de la factura eléctrica, eliminando el sobrecoste por capacidad infrautilizada sin degradar el confort térmico ni la operatividad de la vivienda.

  ---

## Hipótesis 2: Análisis del Consumo Residual y Eficiencia Pasiva
* **Definicion:**  El ratio de consumo base (Standby) frente al consumo nominal es superior al 15% durante los periodos de inactividad o descanso (madrugadas y días no laborables), indicando una ineficiencia en las cargas permanentes.
* **Objetivo:** Cuantificar el impacto económico de las cargas fantasma y sistemas auxiliares (seguridad, recirculación, domótica y modos standby) mediante la comparativa de perfiles de carga entre días laborables y fines de semana.
* **Valor de Negocio:** Identificar si el consumo base es una constante estructural o una ineficiencia operativa. El procesamiento distribuido de Spark nos permitirá segmentar grandes volúmenes de datos temporales para proponer protocolos de "apagado inteligente" o mejoras en el aislamiento eléctrico de circuitos no críticos, reduciendo el gasto energético pasivo de la vivienda.

In [48]:
# H2.1: Analisis de consumos en diferentes periodos

start_h2 = time.time()

periodos_de_consumo = spark.sql("""

SELECT 
    -- 1. Promedio Horas Activas de 6:00 a 00:59 (Active)
    ROUND(AVG(CASE WHEN hour BETWEEN 6 AND 23 OR hour = 0 THEN Global_active_power END), 3) AS AVG_DAY_ACTIVE_KW,
    
    -- 2. Promedio Madrugada de 1:00 a 5:59 (Standby Base)
    ROUND(AVG(CASE WHEN hour BETWEEN 1 AND 5 THEN Global_active_power END), 3) AS AVG_NIGHT_STANDBY_KW,
    
    -- 3. Cálculo del RATIO H2 (Standby vs Active)
    ROUND(
        (AVG(CASE WHEN hour BETWEEN 1 AND 5 THEN Global_active_power END) / 
         AVG(CASE WHEN hour BETWEEN 6 AND 23 OR hour = 0 THEN Global_active_power END)) * 100, 2
    ) AS STANDBY_RATIO_PERCENT,

    -- 4. Comparativa Semanal
    ROUND(AVG(CASE WHEN is_weekend IS FALSE THEN Global_active_power END), 3) AS AVG_WEEKDAY_KW,
    ROUND(AVG(CASE WHEN is_weekend IS TRUE THEN Global_active_power END), 3) AS AVG_WEEKEND_KW,

    -- 5. Cálculo del INCREMENTO (Cuánto sube el consumo el fin de semana vs diario)
    ROUND(
        ((AVG(CASE WHEN is_weekend IS TRUE THEN Global_active_power END) - 
          AVG(CASE WHEN is_weekend IS FALSE THEN Global_active_power END)) / 
         AVG(CASE WHEN is_weekend IS FALSE THEN Global_active_power END)) * 100, 2
    ) AS WEEKEND_INC_PERCENT

FROM power_data;


""")
periodos_de_consumo.show(24)
print(f"H2 procesada en: {time.time() - start_h2:.2f} segundos")


+-----------------+--------------------+---------------------+--------------+--------------+-------------------+
|AVG_DAY_ACTIVE_KW|AVG_NIGHT_STANDBY_KW|STANDBY_RATIO_PERCENT|AVG_WEEKDAY_KW|AVG_WEEKEND_KW|WEEKEND_INC_PERCENT|
+-----------------+--------------------+---------------------+--------------+--------------+-------------------+
|            1.255|               0.472|                37.66|         1.035|         1.234|               19.2|
+-----------------+--------------------+---------------------+--------------+--------------+-------------------+

H2 procesada en: 1.33 segundos


**Análisis Técnico H2.1:**
*  AVG_DAY_ACTIVE (1.255 kW): Es tu línea de base de actividad normal.
*  AVG_NIGHT_STANDBY (0.472 kW): Este es el "gasto fantasma". Para una vivienda, tener medio kilovatio encendido de forma constante toda la noche es el equivalente a tener 50 bombillas LED de 10W o un frigorífico industrial funcionando sin parar.
*  STANDBY_RATIO_PERCENT (37.66%): Casi el 40% de la energía se consume cuando no hay actividad. Indica que la vivienda mantiene una carga activa muy elevada durante la madrugada.
*  WEEKEND_INC_PERCENT: Durante los fines de semana, el consumo medio se incrementa un 19.23% respecto a los días laborables, lo que sugiere una oportunidad de optimización en los hábitos de consumo y programación de equipos térmicos.

  ---

In [49]:
# H2.2: Consumos en la mdrugada

start_h2 = time.time()

consumo_madrugada = spark.sql("""

SELECT 
    'MADRUGADA (01-06h)' AS periodo,
    ROUND(AVG(Global_active_power), 3) as avg_kw_total,
    
    -- Promedios por sub-sistema
    ROUND(AVG(Sub_metering_1 * 0.06), 3) as avg_cocina_kw,
    ROUND(AVG(Sub_metering_2 * 0.06), 3) as avg_lavadero_kw,
    ROUND(AVG(Sub_metering_3 * 0.06), 3) as avg_clima_kw,
    
    -- Desglose porcentual acumulado del bloque
    ROUND((AVG(Sub_metering_1 * 0.06) / AVG(Global_active_power)) * 100, 2) as pct_cocina,
    ROUND((AVG(Sub_metering_2 * 0.06) / AVG(Global_active_power)) * 100, 2) as pct_lavadero,
    ROUND((AVG(Sub_metering_3 * 0.06) / AVG(Global_active_power)) * 100, 2) as pct_clima,
    
    -- El culpable: Consumo Invisible (Otros)
    ROUND(
        ( (AVG(Global_active_power) - (AVG(Sub_metering_1 + Sub_metering_2 + Sub_metering_3) * 0.06)) 
        / AVG(Global_active_power) ) * 100, 2
    ) as pct_otros_invisible

FROM power_data
WHERE hour BETWEEN 1 AND 5;
""")


consumo_madrugada.show(24)
print(f"H2 procesada en: {time.time() - start_h2:.2f} segundos")

+------------------+------------+-------------+---------------+------------+----------+------------+---------+-------------------+
|           periodo|avg_kw_total|avg_cocina_kw|avg_lavadero_kw|avg_clima_kw|pct_cocina|pct_lavadero|pct_clima|pct_otros_invisible|
+------------------+------------+-------------+---------------+------------+----------+------------+---------+-------------------+
|MADRUGADA (01-06h)|       0.472|        0.007|          0.021|       0.121|      1.39|        4.44|    25.58|              68.59|
+------------------+------------+-------------+---------------+------------+----------+------------+---------+-------------------+

H2 procesada en: 1.60 segundos


**Análisis Técnico H2.2:**
*  El desglose de cargas durante la madrugada (01:00-05:59) revela que el 68.59% del consumo residual (0.323 kW) proviene de circuitos no monitorizados (Otros).
*  No son luces, las luces LED modernas no sumarían 323W constantes toda la noche. No son cargadores: Un cargador de móvil consume despreciablemente.
*  Se sospecha de equipos informáticos potentes encendidos (servidores domésticos), frigoríficos/congeladores conectados permanentemente en el circuito de enchufes comunes, o incluso una derivación/pérdida en algún circuito de enchufes generales.


  ---

## Conclusiones del Análisis: Hipótesis 2 (H2)

### **Estado de Validación: ✅ HIPÓTESIS VALIDADA (CRÍTICA)**

Tras el procesamiento distribuido de los perfiles de carga y la segmentación de los periodos de inactividad con Spark SQL, se extraen las siguientes conclusiones técnicas:

* **Confirmación del Consumo Base (Standby):** Se valida la hipótesis con un resultado de **37.66%**, superando drásticamente el umbral del 15% previsto. La vivienda mantiene un consumo constante de **0.472 kW** durante la madrugada (01:00 - 05:59), lo que representa una ineficiencia estructural masiva en el uso de la energía pasiva.
* **Localización del Consumo "Invisible":** El análisis de desglose revela que el **68.59% del consumo residual** proviene de la categoría **"Otros"** (circuitos no monitorizados por Sub-1, 2 o 3). Esto indica que el desperdicio no se debe a electrodomésticos principales, sino a un "goteo" constante en los circuitos de alumbrado general y tomas de corriente (cargas fantasma).
* **Evaluación de la Climatización (Sub-3):** Se descarta la climatización como causa principal del consumo nocturno, ya que solo representa el **25.58%** de la carga en madrugada. Esto sugiere que el aislamiento térmico es aceptable, pero la gestión de cargas electrónicas menores es deficiente.
* **Análisis de Intensidad Semanal:** Se ha detectado un incremento del **19.20%** en el consumo medio durante los fines de semana. Dado que el consumo base (Standby) se mantiene estable, este aumento confirma una falta de protocolos de "apagado inteligente" cuando la actividad humana cesa tras los periodos de mayor uso.



### **Recomendaciones Estratégicas e Impacto en Negocio**

Para corregir esta ineficiencia y reducir el término variable de la factura, se proponen las siguientes acciones:  
1.  **Auditoría y Retrofitting de Iluminación:** Dado que el 70% del desperdicio es "invisible", es imperativo sustituir las luminarias residuales en circuitos generales por tecnología LED de alta eficiencia y verificar el consumo en reposo de equipos IT y servidores domésticos.
2.  **Automatización de Circuitos Críticos (Smart Kill-Switch):** Implementar sistemas de corte automático (relés inteligentes) para los circuitos de entretenimiento y oficina en casa durante la ventana de 01:00 a 06:00. Esta medida permitiría reducir el ratio de standby del 37% a un entorno óptimo del **8-10%**.
3.  **Gestión de la Inercia Térmica:** Reprogramar los sistemas de climatización para realizar un "precargado" térmico antes de la medianoche, evitando arranques cíclicos en la madrugada y aprovechando la envolvente del edificio para mantener el confort sin consumo eléctrico activo.
4.  **Impacto Financiero:** La corrección de este consumo residual de **0.472 kW** a un nivel eficiente de **0.15 kW** supondría un ahorro energético anual de aproximadamente **2,800 kWh**. Esto se traduce en una reducción directa del gasto en energía consumida, complementando el ahorro por potencia contratada logrado en la hipotesis 1 (H1).